### ISSUE TITLE
provide a way to use deep neural networks

### ISSUE THREAD

**Issue by nidhaloff** [nidhaloff/igel#12]
### Description

Igel is built on top of sklearn at the moment. Therefore, All sklearn models can be used. This includes of course the neural network models integrated in sklearn (MLP classifier and MLP regressor). However, sklearn is not powerful enough when it comes to deep neural networks. 
Therefore, this issue aims to include support for using deep neural networks in igel. Maybe Keras API?

### Example:

```
model:
     type: classification # or regression
    algorithm: neural network   # this is already implemented. However, it is using the sklearn NN implementation
    arguments: default   # this will use the default argument of the NN class
```
As you can see. The user can provide these configs in the yaml file and igel will train a neural network. However, the NN model in sklearn is not as powerful as other frameworks like keras, tensorflow, torch etc..

What I mean with this issue and want to implement in the future is **maybe** something like this (feel free to bring new ideas):

```
model:
     deep: true
     type: classification # or regression
    algorithm: neural network   # this will now use the Keras NN model since the user added deep: true 
    arguments: default 
```

OR Maybe even this can be implemented as a VISION ( This will probably take a long time to implement):

```
model:
     uses: keras  # the user can here provide keras, tensorflow or torch 
     type: classification # or regression
    algorithm: neural network   # this will now use the Keras NN model since the user provided that he wants to use keras
    arguments: default 
```

**Comment by dinhanhx** [comment]
I agree with you with using [Keras](https://keras.io/)

Quote from [Why Keras](https://keras.io/why_keras/):
>Keras is an API designed for human beings, not machines. Keras follows best practices for reducing cognitive load: it offers consistent & simple APIs, it minimizes the number of user actions required for common use cases, and it provides clear and actionable feedback upon user error.

This means that implementing deep neural networks for Igel should be easy for Keras users.

**Comment by nidhaloff** [comment]
@dinhanhx The point is not to make using DNN easy for keras users, but instead for everyone. As stated, the goal of this project is to provide machine learning for both technical and non technical users. The user will not interact directly with keras (or whatever framework will be used). I will give you an example, this is a minimal example of what is implemented now:

```
model:
     type: classification # or regression
    algorithm: neural network   # this is already implemented. However, it is using the sklearn NN implementation
    arguments: default   # this will use the default argument of the NN class
```
As you can see. The user can provide these configs in the yaml file and igel will train a neural network. However, the NN model in sklearn is not as powerful as other frameworks like keras, tensorflow, torch etc..

What I mean with this issue and want to implement in the future is maybe something like this:

```
model:
     deep: true
     type: classification # or regression
    algorithm: neural network   # this will now use the Keras NN model since the user added deep: true 
    arguments: default 
```

OR Maybe even this can be implemented as a VISION ( I know this will probably take a long time to implement):

```
model:
     uses: keras  # the user can here provide keras, tensorflow or torch 
     type: classification # or regression
    algorithm: neural network   # this will now use the Keras NN model since the user provided that he wants to use keras
    arguments: default 
```

**Basically, the sky is the limit**

**Comment by dinhanhx** [comment]
I got your idea and motivation of this project. I mentioned "Keras User", I meant that this issue can be resolved by someone familiar with Keras

**Comment by nidhaloff** [comment]
@dinhanhx yes you are right. I was just clarifying for new comers who will read this but I got you ;)

**Comment by Gabriel-Teston** [comment]
Can i make a shot for it? I'm planning on using pytorch and only dense feedforward networks with minimal customization at first.

**Comment by nidhaloff** [comment]
@Gabriel-Teston thanks for joining. I like pytorch but I'm personally not for the idea of using pytorch in igel since more (boilerplate) code will be written and you will spend more time writing code. You can give it a shot sure (at the end of the day nothing will be broken), however, in my opinion some more user friendly framework like **keras**, skorch etc.. will work better for our use case.

**Comment by dinhanhx** [comment]
(I'm back from a Deep learning course with Keras) 
Suggestion: there is an option to choose between using sklearn and keras but [requirements.txt](https://github.com/nidhaloff/igel/blob/master/requirements.txt) stays the same due to the fact that keras is very big. 
- `igel choose-library keras` should __check__ if there is keras or not. If there is not, download and setup keras. After that, make igel use keras as default library for deep learning related models.
- `igel choose-library sklearn` make igel use sklearn completely.
When the user install igel for the first time, it use sklearn completely.
